x/foundation/keeper/msg_server.go

```diff
@@ -297,3 +297,39 @@ func (s msgServer) Revoke(c context.Context, req *foundation.MsgRevoke) (*founda
 
 	return &foundation.MsgRevokeResponse{}, nil
 }
+
+// OneTimeMint defines a method to withdraw coins from the treasury.
+func (s msgServer) OneTimeMint(c context.Context, req *foundation.MsgOneTimeMint) (*foundation.MsgOneTimeMintResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	if err := s.keeper.validateOperator(ctx, req.Operator); err != nil {
+		return nil, err
+	}
+
+	oneTimeMintLeftCount := s.keeper.GetOneTimeMintLeftCount(ctx)
+	if oneTimeMintLeftCount == 0 {
+		return nil, sdkerrors.ErrUnauthorized.Wrapf("The one-time-mint can no longer be executed.")
+	}
+
+	// mint coins to one-time-minter
+	if err := s.keeper.bankKeeper.MintCoins(ctx, foundation.OneTimeMinterName, req.Amount); err != nil {
+		return nil, err
+	}
+
+	// fund treasury from one-time-minter
+	minter := s.keeper.authKeeper.GetModuleAccount(ctx, foundation.OneTimeMinterName).GetAddress()
+	if err := s.keeper.FundTreasury(ctx, minter, req.Amount); err != nil {
+		return nil, err
+	}
+
+	oneTimeMintLeftCount--
+	s.keeper.SetOneTimeMintLeftCount(ctx, oneTimeMintLeftCount)
+
+	if err := ctx.EventManager().EmitTypedEvent(&foundation.EventOneTimeMint{
+		Amount: req.Amount,
+	}); err != nil {
+		panic(err)
```

Comment:
When does this panic occur?
Because if the panic occurs, the `LeftCount` decreased but the mint is fail. Isn't it?

Reply:
All other emit events have the same shape. Under normal circumstances, this cannot happen. Also, Is the failure of the event, the consistency in the store such as left count or mint is always maintained.

Reply:
OK, I understand. Good. Thank you.